cmd/frontend/internal/frontend/metrics.go

```diff
@@ -0,0 +1,260 @@
+/*
+Copyright (c) 2023 Nordix Foundation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package frontend
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+
+	nspAPI "github.com/nordix/meridio/api/nsp/v1"
+	"github.com/nordix/meridio/cmd/frontend/internal/bird"
+	meridioMetrics "github.com/nordix/meridio/pkg/metrics"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+type GatewayMetrics struct {
+	meter            metric.Meter
+	gateways         map[string]*nspAPI.Gateway
+	metricAttributes []metric.ObserveOption
+	RoutingService   *bird.RoutingService
+	mu               sync.Mutex
+}
+
+func NewGatewayMetrics(metricAttributes []metric.ObserveOption) *GatewayMetrics {
+	meter := otel.GetMeterProvider().Meter(meridioMetrics.METER_NAME)
+	gm := &GatewayMetrics{
+		meter:            meter,
+		metricAttributes: metricAttributes,
+	}
+
+	return gm
+}
+
+func (gm *GatewayMetrics) Set(gateways []*nspAPI.Gateway) {
+	gatewaysMap := map[string]*nspAPI.Gateway{}
+
+	for _, gateway := range gateways {
+		gatewaysMap[fmt.Sprintf("NBR-%s", gateway.Name)] = gateway
+	}
+
+	gm.mu.Lock()
+	defer gm.mu.Unlock()
+	gm.gateways = gatewaysMap
+}
+
+// Collect collects the metrics for the gateways.
+func (gm *GatewayMetrics) Collect() error {
+	if gm.RoutingService == nil {
+		return errors.New("routing service not set for gateway metrics")
+	}
+
+	lp, err := gm.RoutingService.LookupCli()
+	if err != nil {
+		return fmt.Errorf("frontend metrics, failed to RoutingService.LookupCli: %w", err)
+	}
+
+	_, err = gm.meter.Int64ObservableCounter(
+		meridioMetrics.MERIDIO_ATTRACTOR_GATEWAY_ROUTES_IMPORTED,
+		metric.WithUnit("routes"),
+		metric.WithDescription("Counts number of routes imported for a gateway."),
+		metric.WithInt64Callback(func(ctx context.Context, observer metric.Int64Observer) error {
+			return gm.observe(
+				ctx,
+				observer,
+				func(birdStats *BirdStats) int64 {
+					return int64(birdStats.routesImported)
```

Comment:
Feels clumsy that the same operation has to be repeated 3 times in the background because 3 values must be extracted. Isn't there some recommendation in opentelemtry how to avoid this?

Reply:
No, I don't think opentelemtry gives any recommendation. they are different types of metrics, like rx/tx bytes/packets for an interface.